### ISSUE TITLE
Eager and Graph API should be more similar

### ISSUE THREAD

**Issue by bionicles** [tensorflow/tensorflow#30763]
**System information**
- TensorFlow version (you are using): 2.0b1
- Are you willing to contribute it (Yes/No): Yes

**Describe the feature and the current behavior/state.**
Eager mode just works. However, the moment you do @tf.function decorators to improve performance, you have to convert a bunch of commands from eager mode into graph mode. 

there are **two ways to do everything in tensorflow: the eager way and the graph way.** This means we have no "single source of truth" (do i use eager which just works but is slow and has memory issues, or do i use graph mode which is lightning fast but requires a ton of extra work to "translate between two dialects" tensorflow eager into tensorflow graph)

**Will this change the current api? How?**
Yes. This will dramatically simplify the tensorflow API, as there will be fewer redundant commands between graph mode and eager mode 

**Who will benefit with this feature?**
Everyone who switches between eager and graph mode would save days of work on every project if this change were implemented

**Any Other info.**
tensorflow 2.0 is a golden opportunity to simplify TF we have less code to debug. I wish the team would merge eager and graph mode as much as possible:

1. tf.shape() vs tensor.shape
2. basic logic often fails in graph mode (for example, if x = "somestring") and there's not many good ways to know why
3. input signature inference would save everyone a lot of time... retracing happens a lot in tf.function without signatures, especially on sequence data with variable shapes, so we have to write a bunch of (None, None ) to fix this ? that's not really informative. this should be automated. Maybe if tf.function sees a shape varying, it can handle this case automatically
4. a nice way to set dtype at the top of the file would be handy
5. a nice way to turn tensorboard tracing on globally would also be handy

```
def trace_function(fn, *args):
    tf.summary.trace_on(graph=True, profiler=True)
    y = fn(*args)
    with writer.as_default():
        tf.summary.trace_export(name="trace_function", step=0, profiler_outdir=log_dir)
    return y
```

**Comment by bionicles** [comment]
in particular, simple logic breaks as soon as you add @tf.function

for example, if you have a multi-task model and the task names are strings, and your eager training loop does
```
if type != "rxn":
```
this stops working in tf.function. Do we need to use some special tensorflow command? Is the string not converted to a tensor? It's so hard to debug this sort of thing because we don't know if it's the data or the API (because there's 2 versions of everything in tensorflow, you have to literally rewrite half your code just to rule out a data problem)

**Comment by bionicles** [comment]
`if step == 0` does not work
`if step < 1` does ...

**Comment by bionicles** [comment]
```
                tf.print(type)
                tf.print(type == 'xyz')
```

> ["xyz"]
> False
Maddening...

**Comment by pandrey-fr** [comment]
Hi,

I totally agree with you on the global message here - it would be really nice indeed to unify things a bit, especially with Eager yielding messy overheads but being rather practical when debugging.

That being said, there currently is a not-so-pleasant solution: do not use tf.function, implement tensorflow control flows instead of using python ones (aka use tf.cond and tf.while_loop), and test your code with `tf.compat.v1.disable_eager_execution()` once it works in Eager. You lose part of the ease that Eager aims at providing, but hey, adding layers of code-translating mechanics is bound to have overheads in addition to the incompatibilities you point out... This is however not perfect because even doing so, I suspect the keras API to embed some code-translation mechanics related to Eager - but at least you can check that your code works without Eager (and probably runs faster too).

**Comment by alextp** [comment]
Sorry, there are a lot of issues raised here, let me tackle them one by one.

> tf.shape() vs tensor.shape

This is tricky. In eager mode we can make shapes always available, but in graph mode it often makes sense to work with partially specified shapes, as this lets you build valid graphs in more circumstances.

Note that working with tf.shape() should give valid results in both eager and graph modes (as we can constant-fold it away if necessary).

> basic logic often fails in graph mode (for example, if x = "somestring") and there's not many good ways to know why

In both eager and graph modes tensor equality (operator ==, not the operator = which is not an operator in python) is not overridden. This is due to legacy reasons, and we're trying to fix it in TF2, but it's not a difference between eager and graph modes.

That said you rarely need to manipulate strings using tensors of strings, so I don't know when you're hitting this with string dtypes.

> input signature inference would save everyone a lot of time... retracing happens a lot in tf.function without signatures, especially on sequence data with variable shapes, so we have to write a bunch of (None, None ) to fix this ? that's not really informative. this should be automated. Maybe if tf.function sees a shape varying, it can handle this case automatically

We're working on improving this. You can try passing experimental_relax_shapes=True to tf.function, and then we try to infer the signature for you. If you encounter any problems when using that please file issues as we'd like to improve it to the point where we can remove the experimental bit and turn it on by default.

> a nice way to set dtype at the top of the file would be handy

We're working on this as part of the mixed-precision APIs. @reedwm for FYI.

> a nice way to turn tensorboard tracing on globally would also be handy

Please file a separate issue for this.

**Comment by bionicles** [comment]
Thank you @alextp @pandrey-fr

**Comment by ravikyram** [comment]
@bionicles 
Can we close the issue since the query is been answered. Please,let us know. Thanks!

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this still an issue?

**Comment by jvishnuvardhan** [comment]
Automatically closing this out since I understand it to be resolved, but please let me know if I'm mistaken.Thanks!
